### ISSUE TITLE
Flow errors 

### ISSUE THREAD

**Issue by skovhus** [invertase/react-native-firebase#834]
### Issue

With a standard React Native .flowconfig setup using 0.63.0 and React Native 0.53.0, I'm getting 25 errors after installing react-native-firebase:

```
$ flow --show-all-errors --quiet
Error: node_modules/react-native-firebase/dist/modules/auth/PhoneAuthListener.js.flow:113
113:         this[`_${type}Handler`].bind(this)
             ^^^^^^^^^^^^^^^^^^^^^^^ computed property. Indexable signature not found in
113:         this[`_${type}Handler`].bind(this)
             ^^^^ PhoneAuthListener

Error: node_modules/react-native-firebase/dist/modules/core/firebase-app.js.flow:133
133:       this[key] = props[key];
           ^^^^^^^^^ assignment of computed property/element. Indexable signature not found in
133:       this[key] = props[key];
           ^^^^ App

Error: node_modules/react-native-firebase/dist/modules/database/reference.js.flow:317
317:           .then(() => onComplete(null, newRef))
                           ^^^^^^^^^^^^^^^^^^^^^^^^ call of `onComplete`. Function cannot be called on possibly undefined value
317:           .then(() => onComplete(null, newRef))
                           ^^^^^^^^^^ undefined

Error: node_modules/react-native-firebase/dist/modules/database/reference.js.flow:319
319:           .catch(error => onComplete(error, null))
                               ^^^^^^^^^^^^^^^^^^^^^^^ call of `onComplete`. Function cannot be called on possibly undefined value
319:           .catch(error => onComplete(error, null))
                               ^^^^^^^^^^ undefined

Error: node_modules/react-native-firebase/dist/modules/firestore/DocumentReference.js.flow:103
                      v
103:       observer = {
104:         next: optionsOrObserverOrOnNext,
105:         error: observerOrOnNextOrOnError,
106:       };
           ^ object literal. This type is incompatible with
 93:     let observer: Observer;
                       ^^^^^^^^ Observer
  Property `error` is incompatible:
    105:         error: observerOrOnNextOrOnError,
                        ^^^^^^^^^^^^^^^^^^^^^^^^^ Observer. This type is incompatible with
     30:   error?: ObserverOnError,
                   ^^^^^^^^^^^^^^^ ObserverOnError
      Callable property is incompatible:
         30:   error?: ObserverOnError,
                       ^^^^^^^^^^^^^^^ ObserverOnError. Callable signature not found in
        105:         error: observerOrOnNextOrOnError,
                            ^^^^^^^^^^^^^^^^^^^^^^^^^ Observer

Error: node_modules/react-native-firebase/dist/modules/firestore/DocumentReference.js.flow:103
                      v
103:       observer = {
104:         next: optionsOrObserverOrOnNext,
105:         error: observerOrOnNextOrOnError,
106:       };
           ^ object literal. This type is incompatible with
 93:     let observer: Observer;
                       ^^^^^^^^ Observer
  Property `next` is incompatible:
    104:         next: optionsOrObserverOrOnNext,
                       ^^^^^^^^^^^^^^^^^^^^^^^^^ DocumentListenOptions. This type is incompatible with
     31:   next: ObserverOnNext,
                 ^^^^^^^^^^^^^^ ObserverOnNext
      Callable property is incompatible:
         31:   next: ObserverOnNext,
                     ^^^^^^^^^^^^^^ ObserverOnNext. Callable signature not found in
        104:         next: optionsOrObserverOrOnNext,
                           ^^^^^^^^^^^^^^^^^^^^^^^^^ DocumentListenOptions

Error: node_modules/react-native-firebase/dist/modules/firestore/DocumentReference.js.flow:103
                      v
103:       observer = {
104:         next: optionsOrObserverOrOnNext,
105:         error: observerOrOnNextOrOnError,
106:       };
           ^ object literal. This type is incompatible with
 93:     let observer: Observer;
                       ^^^^^^^^ Observer
  Property `next` is incompatible:
    104:         next: optionsOrObserverOrOnNext,
                       ^^^^^^^^^^^^^^^^^^^^^^^^^ Observer. This type is incompatible with
     31:   next: ObserverOnNext,
                 ^^^^^^^^^^^^^^ ObserverOnNext
      Callable property is incompatible:
         31:   next: ObserverOnNext,
                     ^^^^^^^^^^^^^^ ObserverOnNext. Callable signature not found in
        104:         next: optionsOrObserverOrOnNext,
                           ^^^^^^^^^^^^^^^^^^^^^^^^^ Observer

Error: node_modules/react-native-firebase/dist/modules/firestore/DocumentReference.js.flow:123
                          v
123:           observer = {
124:             next: optionsOrObserverOrOnNext.next,
125:             error: optionsOrObserverOrOnNext.error,
126:           };
               ^ object literal. This type is incompatible with
 93:     let observer: Observer;
                       ^^^^^^^^ Observer
  Property `error` is incompatible:
    125:             error: optionsOrObserverOrOnNext.error,
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ property `error` of unknown type. This type is incompatible with
     30:   error?: ObserverOnError,
                   ^^^^^^^^^^^^^^^ ObserverOnError

Error: node_modules/react-native-firebase/dist/modules/firestore/DocumentReference.js.flow:123
                          v
123:           observer = {
124:             next: optionsOrObserverOrOnNext.next,
125:             error: optionsOrObserverOrOnNext.error,
126:           };
               ^ object literal. This type is incompatible with
 93:     let observer: Observer;
                       ^^^^^^^^ Observer
  Property `next` is incompatible:
    124:             next: optionsOrObserverOrOnNext.next,
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ property `next` of unknown type. This type is incompatible with
     31:   next: ObserverOnNext,
                 ^^^^^^^^^^^^^^ ObserverOnNext

Error: node_modules/react-native-firebase/dist/modules/firestore/DocumentReference.js.flow:147
                          v
147:           observer = {
148:             next: observerOrOnNextOrOnError,
149:             error: onError,
150:           };
               ^ object literal. This type is incompatible with
 93:     let observer: Observer;
                       ^^^^^^^^ Observer
  Property `next` is incompatible:
    148:             next: observerOrOnNextOrOnError,
                           ^^^^^^^^^^^^^^^^^^^^^^^^^ Observer. This type is incompatible with
     31:   next: ObserverOnNext,
                 ^^^^^^^^^^^^^^ ObserverOnNext
      Callable property is incompatible:
         31:   next: ObserverOnNext,
                     ^^^^^^^^^^^^^^ ObserverOnNext. Callable signature not found in
        148:             next: observerOrOnNextOrOnError,
                               ^^^^^^^^^^^^^^^^^^^^^^^^^ Observer

Error: node_modules/react-native-firebase/dist/modules/firestore/DocumentReference.js.flow:147
                          v
147:           observer = {
148:             next: observerOrOnNextOrOnError,
149:             error: onError,
150:           };
               ^ object literal. This type is incompatible with
 93:     let observer: Observer;
                       ^^^^^^^^ Observer
  Property `next` is incompatible:
    148:             next: observerOrOnNextOrOnError,
                           ^^^^^^^^^^^^^^^^^^^^^^^^^ undefined. This type is incompatible with
     31:   next: ObserverOnNext,
                 ^^^^^^^^^^^^^^ ObserverOnNext

Error: node_modules/react-native-firebase/dist/modules/firestore/Query.js.flow:191
                      v
191:       observer = {
192:         next: optionsOrObserverOrOnNext,
193:         error: observerOrOnNextOrOnError,
194:       };
           ^ object literal. This type is incompatible with
181:     let observer: Observer;
                       ^^^^^^^^ Observer
  Property `error` is incompatible:
    193:         error: observerOrOnNextOrOnError,
                        ^^^^^^^^^^^^^^^^^^^^^^^^^ Observer. This type is incompatible with
     70:   error?: ObserverOnError,
                   ^^^^^^^^^^^^^^^ ObserverOnError
      Callable property is incompatible:
         70:   error?: ObserverOnError,
                       ^^^^^^^^^^^^^^^ ObserverOnError. Callable signature not found in
        193:         error: observerOrOnNextOrOnError,
                            ^^^^^^^^^^^^^^^^^^^^^^^^^ Observer

Error: node_modules/react-native-firebase/dist/modules/firestore/Query.js.flow:191
                      v
191:       observer = {
192:         next: optionsOrObserverOrOnNext,
193:         error: observerOrOnNextOrOnError,
194:       };
           ^ object literal. This type is incompatible with
181:     let observer: Observer;
                       ^^^^^^^^ Observer
  Property `next` is incompatible:
    192:         next: optionsOrObserverOrOnNext,
                       ^^^^^^^^^^^^^^^^^^^^^^^^^ Observer. This type is incompatible with
     71:   next: ObserverOnNext,
                 ^^^^^^^^^^^^^^ ObserverOnNext
      Callable property is incompatible:
         71:   next: ObserverOnNext,
                     ^^^^^^^^^^^^^^ ObserverOnNext. Callable signature not found in
        192:         next: optionsOrObserverOrOnNext,
                           ^^^^^^^^^^^^^^^^^^^^^^^^^ Observer

Error: node_modules/react-native-firebase/dist/modules/firestore/Query.js.flow:191
                      v
191:       observer = {
192:         next: optionsOrObserverOrOnNext,
193:         error: observerOrOnNextOrOnError,
194:       };
           ^ object literal. This type is incompatible with
181:     let observer: Observer;
                       ^^^^^^^^ Observer
  Property `next` is incompatible:
    192:         next: optionsOrObserverOrOnNext,
                       ^^^^^^^^^^^^^^^^^^^^^^^^^ QueryListenOptions. This type is incompatible with
     71:   next: ObserverOnNext,
                 ^^^^^^^^^^^^^^ ObserverOnNext
      Callable property is incompatible:
         71:   next: ObserverOnNext,
                     ^^^^^^^^^^^^^^ ObserverOnNext. Callable signature not found in
        192:         next: optionsOrObserverOrOnNext,
                           ^^^^^^^^^^^^^^^^^^^^^^^^^ QueryListenOptions

Error: node_modules/react-native-firebase/dist/modules/firestore/Query.js.flow:239
                          v
239:           observer = {
240:             next: observerOrOnNextOrOnError,
241:             error: onError,
242:           };
               ^ object literal. This type is incompatible with
181:     let observer: Observer;
                       ^^^^^^^^ Observer
  Property `next` is incompatible:
    240:             next: observerOrOnNextOrOnError,
                           ^^^^^^^^^^^^^^^^^^^^^^^^^ Observer. This type is incompatible with
     71:   next: ObserverOnNext,
                 ^^^^^^^^^^^^^^ ObserverOnNext
      Callable property is incompatible:
         71:   next: ObserverOnNext,
                     ^^^^^^^^^^^^^^ ObserverOnNext. Callable signature not found in
        240:             next: observerOrOnNextOrOnError,
                               ^^^^^^^^^^^^^^^^^^^^^^^^^ Observer

Error: node_modules/react-native-firebase/dist/modules/firestore/Query.js.flow:239
                          v
239:           observer = {
240:             next: observerOrOnNextOrOnError,
241:             error: onError,
242:           };
               ^ object literal. This type is incompatible with
181:     let observer: Observer;
                       ^^^^^^^^ Observer
  Property `next` is incompatible:
    240:             next: observerOrOnNextOrOnError,
                           ^^^^^^^^^^^^^^^^^^^^^^^^^ undefined. This type is incompatible with
     71:   next: ObserverOnNext,
                 ^^^^^^^^^^^^^^ ObserverOnNext

Error: node_modules/react-native-firebase/dist/utils/apps.js.flow:36
 36:     return Object.values(APPS);
                ^^^^^^^^^^^^^^^^^^^ array type. This type is incompatible with the expected return type of
 34:   apps(): Array<App> {
               ^^^^^^^^^^ array type
  Type argument `T` is incompatible:
     49:     static values(object: any): Array<mixed>;
                                               ^^^^^ mixed. This type is incompatible with. See lib: /private/tmp/flow/flowlib_24c3a570/core.js:49
     34:   apps(): Array<App> {
                         ^^^ App

Error: node_modules/react-native-firebase/dist/utils/apps.js.flow:178
178:       const module = _app[namespace];
                          ^^^^^^^^^^^^^^^ computed property. Indexable signature not found in
178:       const module = _app[namespace];
                          ^^^^ App

Error: node_modules/react-native-firebase/dist/utils/internals.js.flow:25
       v---
 25:   1: {
 26:     code: 'SERVICE_MISSING',
 27:     message: 'Google Play services is missing on this device.',
 28:   },
       ^ non-string literal property keys not supported

Error: node_modules/react-native-firebase/dist/utils/internals.js.flow:30
       v---
 30:   2: {
 31:     code: 'SERVICE_VERSION_UPDATE_REQUIRED',
 32:     message:
 33:       'The installed version of Google Play services on this device is out of date.',
 34:   },
       ^ non-string literal property keys not supported

Error: node_modules/react-native-firebase/dist/utils/internals.js.flow:36
       v---
 36:   3: {
 37:     code: 'SERVICE_DISABLED',
 38:     message:
 39:       'The installed version of Google Play services has been disabled on this device.',
 40:   },
       ^ non-string literal property keys not supported

Error: node_modules/react-native-firebase/dist/utils/internals.js.flow:42
       v---
 42:   9: {
 43:     code: 'SERVICE_INVALID',
 44:     message:
 45:       'The version of the Google Play services installed on this device is not authentic.',
 46:   },
       ^ non-string literal property keys not supported

Error: node_modules/react-native-firebase/dist/utils/internals.js.flow:48
       v----
 48:   18: {
 49:     code: 'SERVICE_UPDATING',
 50:     message: 'Google Play services is currently being updated on this device.',
 51:   },
       ^ non-string literal property keys not supported

Error: node_modules/react-native-firebase/dist/utils/internals.js.flow:53
       v----
 53:   19: {
 54:     code: 'SERVICE_MISSING_PERMISSION',
 55:     message:
 56:       "Google Play service doesn't have one or more required permissions.",
 57:   },
       ^ non-string literal property keys not supported

Error: node_modules/react-native-firebase/lib/modules/core/firebase.js:0
Firebase. Duplicate module provider
current provider. See: node_modules/react-native-firebase/dist/modules/core/firebase.js:0

Found 25 errors
```

Flow config file for React Native can be found here https://raw.githubusercontent.com/ncuillery/rn-diff/a943ac81415eee364219d3cac3afc0b5ab0a9a75/RnDiffApp/.flowconfig

### Environment

n/a (this is a flow issue

2. Development Operating System:

node: v9.2.1
flow: 0.63.0
mac OS Sierra

3. Build Tools:

n/a

4. React Native version:

0.53.0

5. RNFirebase Version:

n/a

6. Firebase Module:

3.2.4

**Comment by skovhus** [comment]
My workaround for now is to add `.*/node_modules/react-native-firebase` to the `[ignore]` section in my `.flowconfig`... But clearly I'm losing all type information, so this is just a temporary fix.

**Comment by chrisbianca** [comment]
Can you try adding the following line to your .flowconfig with the other suppress_comment lines:

```
suppress_comment=\\(.\\|\n\\)*\\$FlowBug.*
```

**Comment by skovhus** [comment]
After doing that I'm just getting:
```
Error: node_modules/react-native-firebase/lib/modules/core/firebase.js:0
Firebase. Duplicate module provider
current provider. See: node_modules/react-native-firebase/dist/modules/core/firebase.js:0
```

After excluding `.*/node_modules/react-native-firebase/lib` in the `[ignore]` section. I only get the `String` issue I fixed in #832

**Comment by chrisbianca** [comment]
Thanks for the update. That's good - just means we have to use one of the more standard flow ignore markers.  I assumed that the flowconfig from the module would be followed, but clearly not.

The duplicate module error is a strange one - do you have any other modules related to Firebase that might also be marking itself as a Firebase modules? Specifically:

```
@providesModule Firebase
```

**Comment by skovhus** [comment]
Actually this isn’t weird to me. Both the dist and lib folder provides this module. 

Two options that I can think of:
1) make people update their flowconfig (not that nice)
2) exclude lib folder from the published package

**Comment by matt-oakes** [comment]
Option 2 is what should be happening anyway to keep the module smaller.

**Comment by skovhus** [comment]
Also makes sense IMO. : )

**Comment by chrisbianca** [comment]
The custom $FlowBug tag has been replaced as part of this commit: https://github.com/invertase/react-native-firebase/commit/c57c6a74a848fafead8bc838e28bf56c29dd6a54

In terms of excluding lib, as explained here: https://github.com/invertase/react-native-firebase/pull/835#issuecomment-366441229, this doesn't work for us because we want to also be able to install dynamically from specific github revisions and branches.  If you have any other suggestions for how to get this working, we'd be happy to hear them...

Instead, I have just removed the provideModule declaration as it looks like this is a hangover from an older version of RN that's no longer required.

I'll close this, as the fixes will be part of our next release at some point over the next few days.
